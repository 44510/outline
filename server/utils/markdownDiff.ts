import { findIndex, findLastIndex } from "lodash";
import diff from "node-htmldiff";
import renderToHtml from "@server/editor/renderToHtml";

export default function markdownDiff(
  before: string,
  after: string,
  fullDiff = false,
  buffer = 1
) {
  // The basic idea here is to first render the Markdown to HTML, then diff the
  // HTML - both sides will have valid HTML so we should have a valid diff as well

  const beforeHtml = renderToHtml(before);
  const afterHtml = renderToHtml(after);
  const diffHtml = diff(beforeHtml, afterHtml);

  if (fullDiff) {
    return diffHtml;
  }

  if (before === after) {
    return "";
  }

  // Split diff at paragraphs and find the first and last changed tags
  // so we can chop around paragraphs rather than return the entire document.
  //
  // In an ideal world we'd use an AST here and parse that rather than be doing
  // operations on strings. I hope this can be revisted in the future with an
  // improved diffing library.
  const newParagraph = /(?:^|\n)<p>/;
  let lines = diffHtml.split(newParagraph);

  const firstChangedLineIndex = findIndex(
    lines,
    (value) => value.includes("<ins ") || value.includes("<del ")
  );
  const lastChangedLineIndex = findLastIndex(
    lines,
    (value) => value.includes("</ins>") || value.includes("</del>")
  );

  const start = Math.max(0, firstChangedLineIndex - buffer);
  const end = Math.min(lines.length, lastChangedLineIndex + buffer);
  lines = lines.slice(start, end);

  if (!lines.length) {
    return "";
  }

  return [start > 0 ? "" : undefined, ...lines]
    .filter((x) => x !== undefined)
    .join("\n<p>")
    .trim();
}
