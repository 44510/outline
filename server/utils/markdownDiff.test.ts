import fs from "fs";
import path from "path";
import markdownDiff from "./markdownDiff";

it("should diff a complex document", async () => {
  const before = await fs.promises.readFile(
    path.resolve(process.cwd(), "server", "test", "fixtures", "complex.md"),
    "utf8"
  );

  const after = await fs.promises.readFile(
    path.resolve(
      process.cwd(),
      "server",
      "test",
      "fixtures",
      "complexModified.md"
    ),
    "utf8"
  );

  const diff = markdownDiff(before, after);
  expect(diff).toMatchSnapshot();
});

it("should return empty string when both sides are empty", () => {
  const diff = markdownDiff("", "");
  expect(diff).toEqual("");
});

it("should return everything inserted when previously empty", async () => {
  const content = await fs.promises.readFile(
    path.resolve(process.cwd(), "server", "test", "fixtures", "complex.md"),
    "utf8"
  );

  const diff = markdownDiff("", content);
  expect(diff).toMatchSnapshot();
});

it("should return empty for changed nodes", async () => {
  // Note: This isn't ideal behavior, but it is current behavior. If the diffing
  // library is improved then we could potentially render the old + new heading
  // with ins/del tags as appropriate.
  const diff = markdownDiff("# Heading", "## Heading");
  expect(diff).toEqual("");
});

it("should return deleted nodes", async () => {
  const diff = markdownDiff("![caption](/image.png)", "");
  expect(diff).toEqual(
    '<p><del data-operation-index="0"><img src="/image.png" alt="caption"></del></p>'
  );
});
