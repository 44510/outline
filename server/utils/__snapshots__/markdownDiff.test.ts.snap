// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should diff a complex document 1`] = `
"<p>This is a second test paragraph. This is a second sentence.</p>
<p>This is a another test paragraph. This is a another sentence.</p>
<ul>
<li>list item 1</li>
<li data-diff-node=\\"del\\" data-operation-index=\\"1\\"><del data-operation-index=\\"1\\">list item 2</del></li></ul>
<pre><code>this is a codeblock
</code></pre><p data-diff-node=\\"ins\\" data-operation-index=\\"3\\"><ins data-operation-index=\\"3\\">This is a new paragraph.</ins></p>
<div class=\\"notice notice-info\\">
<p>This is an info block</p>
</div>
<p><span class=\\"placeholder\\">This is a placeholder</span></p>
<p><span class=\\"highlight\\">this is a highlight</span></p>
<ul>
<li class=\\"checkbox-list-item\\"><span class=\\"checkbox checked\\">[<del data-operation-index=\\"5\\"> ]</del><ins data-operation-index=\\"5\\">x]</ins></span>checklist item 1</li><li class=\\"checkbox-list-item\\"><span class=\\"checkbox checked\\" data-diff-node=\\"ins\\" data-operation-index=\\"7\\"><ins data-operation-index=\\"7\\">[x]</ins></span><ins data-operation-index=\\"7\\">checklist item 2</ins></li>
<li class=\\"checkbox-list-item\\"><span class=\\"checkbox \\">[ ]</span>checklist item <del data-operation-index=\\"9\\">2</del><ins data-operation-index=\\"9\\">3</ins></li><li class=\\"checkbox-list-item\\"><span class=\\"checkbox \\" data-diff-node=\\"ins\\" data-operation-index=\\"9\\"><ins data-operation-index=\\"9\\">[ ]</ins></span><ins data-operation-index=\\"9\\">checklist item 4</ins></li>
<li class=\\"checkbox-list-item\\"><span class=\\"checkbox checked\\">[x]</span>checklist item <del data-operation-index=\\"11\\">3</del><ins data-operation-index=\\"11\\">5</ins></li>
</ul>"
`;

exports[`should return everything inserted when previously empty 1`] = `
"<h1 data-diff-node=\\"ins\\" data-operation-index=\\"0\\"><ins data-operation-index=\\"0\\">Heading 1</ins></h1><h2 data-diff-node=\\"ins\\" data-operation-index=\\"0\\"><ins data-operation-index=\\"0\\">Heading 2</ins></h2><p data-diff-node=\\"ins\\" data-operation-index=\\"0\\"><ins data-operation-index=\\"0\\">This is a test paragraph</ins></p><p data-diff-node=\\"ins\\" data-operation-index=\\"0\\"><ins data-operation-index=\\"0\\">This is a second test paragraph. This is a second sentence.</ins></p><p data-diff-node=\\"ins\\" data-operation-index=\\"0\\"><ins data-operation-index=\\"0\\">This is a another test paragraph. This is a another sentence.</ins></p><ul data-diff-node=\\"ins\\" data-operation-index=\\"0\\"><li data-diff-node=\\"ins\\" data-operation-index=\\"0\\"><ins data-operation-index=\\"0\\">list item 1</ins></li><li data-diff-node=\\"ins\\" data-operation-index=\\"0\\"><ins data-operation-index=\\"0\\">list item 2</ins></li></ul><pre data-diff-node=\\"ins\\" data-operation-index=\\"0\\"><code data-diff-node=\\"ins\\" data-operation-index=\\"0\\"><ins data-operation-index=\\"0\\">this is a codeblock
</ins></code></pre><div class=\\"notice notice-info\\" data-diff-node=\\"ins\\" data-operation-index=\\"0\\"><p data-diff-node=\\"ins\\" data-operation-index=\\"0\\"><ins data-operation-index=\\"0\\">This is an info block</ins></p></div><p data-diff-node=\\"ins\\" data-operation-index=\\"0\\"><span class=\\"placeholder\\" data-diff-node=\\"ins\\" data-operation-index=\\"0\\"><ins data-operation-index=\\"0\\">This is a placeholder</ins></span></p><p data-diff-node=\\"ins\\" data-operation-index=\\"0\\"><span class=\\"highlight\\" data-diff-node=\\"ins\\" data-operation-index=\\"0\\"><ins data-operation-index=\\"0\\">this is a highlight</ins></span></p><ul data-diff-node=\\"ins\\" data-operation-index=\\"0\\"><li class=\\"checkbox-list-item\\" data-diff-node=\\"ins\\" data-operation-index=\\"0\\"><span class=\\"checkbox \\" data-diff-node=\\"ins\\" data-operation-index=\\"0\\"><ins data-operation-index=\\"0\\">[ ]</ins></span><ins data-operation-index=\\"0\\">checklist item 1</ins></li><li class=\\"checkbox-list-item\\" data-diff-node=\\"ins\\" data-operation-index=\\"0\\"><span class=\\"checkbox \\" data-diff-node=\\"ins\\" data-operation-index=\\"0\\"><ins data-operation-index=\\"0\\">[ ]</ins></span><ins data-operation-index=\\"0\\">checklist item 2</ins></li><li class=\\"checkbox-list-item\\" data-diff-node=\\"ins\\" data-operation-index=\\"0\\"><span class=\\"checkbox checked\\" data-diff-node=\\"ins\\" data-operation-index=\\"0\\"><ins data-operation-index=\\"0\\">[x]</ins></span><ins data-operation-index=\\"0\\">checklist item 3</ins></li></ul><p data-diff-node=\\"ins\\" data-operation-index=\\"0\\"><ins data-operation-index=\\"0\\">same on both sides</ins></p><p data-diff-node=\\"ins\\" data-operation-index=\\"0\\"><ins data-operation-index=\\"0\\">same on both sides</ins></p><p data-diff-node=\\"ins\\" data-operation-index=\\"0\\"><ins data-operation-index=\\"0\\">same on both sides</ins></p>"
`;
